csrc/codebase/mmpose/keypoints_from_heatmap.cpp

```diff
@@ -0,0 +1,396 @@
+                     (int)heatmap.data_type());
+      return Status(eNotSupported);
+    }
+
+    auto& img_metas = _data["img_metas"];
+
+    vector<float> center;
+    vector<float> scale;
+    from_value(img_metas["center"], center);
+    from_value(img_metas["scale"], scale);
+    Tensor pred =
+        keypoints_from_heatmap(heatmap, center, scale, unbiased_decoding_, post_process_,
+                               modulate_kernel_, valid_radius_factor_, use_udp_, target_type_);
+
+    return GetOutput(pred);
+  }
+
+  Value GetOutput(Tensor& pred) {
+    PoseDetectorOutput output;
+    int K = pred.shape(1);
+    float* data = pred.data<float>();
+    for (int i = 0; i < K; i++) {
+      float x = *(data + 0);
+      float y = *(data + 1);
+      float s = *(data + 2);
+      output.key_points.push_back({{x, y}, s});
+      data += 3;
+    }
+    return to_value(std::move(output));
+  }
+
+  Tensor keypoints_from_heatmap(const Tensor& _heatmap, const vector<float>& center,
+                                const vector<float>& scale, bool unbiased_decoding,
+                                const string& post_process, int modulate_kernel,
+                                float valid_radius_factor, bool use_udp,
+                                const string& target_type) {
+    Tensor heatmap(_heatmap.desc());
+    heatmap.CopyFrom(_heatmap, stream()).value();
+    stream().Wait().value();
+
+    int K = heatmap.shape(1);
+    int H = heatmap.shape(2);
+    int W = heatmap.shape(3);
+
+    if (post_process == "megvii") {
+      heatmap = gaussian_blur(heatmap, modulate_kernel);
+    }
+
+    Tensor pred;
+
+    if (use_udp) {
+      if (to_lower(target_type) == to_lower(string("GaussianHeatMap"))) {
+        pred = get_max_pred(heatmap);
+        post_dark_udp(pred, heatmap, modulate_kernel);
+      } else if (to_lower(target_type) == to_lower(string("CombinedTarget"))) {
+        // output channel = 3 * channel_cfg['num_output_channels']
+        assert(K % 3 == 0);
+        cv::parallel_for_(cv::Range(0, K), [&](const cv::Range& r) {
+          for (int i = r.start; i < r.end; i++) {
+            int kt = (i % 3 == 0) ? 2 * modulate_kernel + 1 : modulate_kernel;
+            float* data = heatmap.data<float>() + i * H * W;
+            cv::Mat work = cv::Mat(H, W, CV_32FC(1), data);
+            cv::GaussianBlur(work, work, {kt, kt}, 0);  // inplace
+          }
+        });
+        float valid_radius = valid_radius_factor_ * H;
+        TensorDesc desc = {Device{"cpu"}, {DataType::kFLOAT}, {1, K / 3, H, W}};
+        Tensor offset_x(desc);
+        Tensor offset_y(desc);
+        Tensor heatmap_(desc);
+        {
+          // split heatmap
+          float* src = heatmap.data<float>();
+          float* dst0 = heatmap_.data<float>();
+          float* dst1 = offset_x.data<float>();
+          float* dst2 = offset_y.data<float>();
+          for (int i = 0; i < K; i++) {
```

Comment:
I think we can simplify this loop, removing if-elseif-else branch
```C++
for (i = 0; i < K / 3; ++i) {
    // 1. [src, src+H*W) to dst0;
    // 2. [src+H*W, src+2*H*W) to dst1;
    // 3. [src+2*H*W, src+3*H*W) to dst2;
}
```

Reply:
done